shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm

```diff
@@ -171,7 +163,20 @@ - (BOOL)application:(UIApplication*)application
   if ([_lifeCycleDelegate application:application openURL:url options:options]) {
     return YES;
   }
-  return [self openURL:url];
+
+  __block BOOL openURLSuccess = NO;  // Flag to track success
+  // Use a dispatch semaphore to wait for the completion handler
+  dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
```

Comment:
You won't want a dispatch_semaphore on the main thread, that will cause hangs and potentially deadlocks.  

The [run loop](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Multithreading/RunLoopManagement/RunLoopManagement.html) is a processing loop for event handling, and there's one assigned to the main thread.  I believe what you want is to kick off the request, and spin the main run loop until it completes, then return the value from this method.

I didn't run this code so you should test it, but something like:

```objc
  __block BOOL openURLSuccess = NO;
  __block BOOL openURLCompleted = NO;
  [self openURL:url
                options:options
      completionHandler:^(BOOL success) {
        openURLSuccess = success;
        openURLCompleted = YES;
      }];

  while (!openURLCompleted) {
    [NSRunLoop.currentRunLoop runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
  }
  return openURLSuccess;
```

Reply:
Thanks for educating me about the The [run loop](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Multithreading/RunLoopManagement/RunLoopManagement.html) 

I have some more questions : 
1.  is a while loop a form of busy waiting and will it cause some performance issues? Should i also add a timeout to prevent an infinite loop?
3.  Looks like the async api [openURL:options:completionHandler:] is newer than the non-async api [application:openURL:options:] , is it possible the async one is replacing the non-async one?  if so,  it may be ok the async api returns the value from the framework but the old non-async api remains the old behavior to return the flag value.

Reply:
> is a while loop a form of busy waiting and will it cause some performance issues? Should i also add a timeout to prevent an infinite loop?

Are you concerned the code you're calling `didTimeout` logic won't work? You could pass in that timeout into the view controller method so it's obvious it handles timing out?

Reply:
I'm also a bit concerned  " navigationChannel invokeMethod:arguments:result: " will time out  😅  Is it possible?

Reply:
You could add a timeout instead of the while loop, that would work too. And return NO if it times out.  But if it times out does that mean you need to cancel the `pushRouteInformation` request?

Reply:
@hangyujin To add a timeout, you can do this: 

```
  CFTimeInterval start = CACurrentMediaTime();
  [self openURL...:{
    // ...
  }];
  while (!openURLCompleted && CACurrentMediaTime() - start <= THE_TIME_OUT_IN_SECONDS) {
    [NSRunLoop.currentRunLoop runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
}

Reply:
I added a timeout here.
'''
it would be bad if it times out but the pushRouteInformation request has gone through, but not yet returned?
'''

1. Give the request a timeout ---> The timeout may be too short, causing the openURL method to return NO early
2. Don't give the request a timeout ----> It may cause an infinite loop,

Can I set a slightly longer timeout (maybe 7 seconds) to mediate 1 and 2?

Reply:
From the code, the delay comes from `waitForFirstFrame` (3 seconds) + `pushRouteInformation` (unknown delay). I'd run it and see how long it takes `pushRouteInformation` to call its `result` block (you can do so by printing out the time elapsed), then pick the shortest delay that you feel confident that will cover `pushRouteInformation`.

Reply:
> is a while loop a form of busy waiting and will it cause some performance issues? 

From the API doc the UI shouldn't be blocked, but please test it out. It will be easier to find out by adding a dummy delay to your completion block, so you have enough time to scroll the screen after the open URL call. Something like: 

```
  [self openURL:url
                options:options
      completionHandler:^(BOOL success) {

          // Note: delay this call for 20 seconds for debugging purpose
          dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(20 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{

            openURLSuccess = success;
            openURLCompleted = YES;
        }
      }];
```

Reply:
Sorry for the delay, i just got my iOS device registered under Google's Apple developer account so i can test deep link on it.

1. i tested pushRouteInformation' ant it's under a second.  i think 5 seconds in total can cover it.
2. you're correct, adding delay will not cause UI to be blocked.
3. but it looks like my run loop will have a deadlock with the dispatch_async here: https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L1428  https://github.com/flutter/engine/blob/85f44297d957b7282b804754ded778667ea18a89/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L1428 because the dispatch_async waits for runloop to end and the  runloop is waiting for the callback from dispatch_async. So I don't think i should use a run loop.

Reply:
Let me try it out

Reply:
basically it's like 
```
- (void)fakedispatch:(void (^)(BOOL success))completion {
  dispatch_queue_t queue = dispatch_get_main_queue();
  dispatch_async(queue, ^{
     NSLog(@"fakedispatch:completionHandler: "); 
      completion(YES);
  });
}

- (void)fakeRunloop {
  __block BOOL completed = NO;
  CFTimeInterval startTime = CACurrentMediaTime();
  [self fakedispatch:^(BOOL success) {
    completed = YES;
  }];
  while (!completed) {
    NSLog(@"runing loop %f", CACurrentMediaTime() - startTime);
    [NSRunLoop.currentRunLoop runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1.0]];
  }
}
```
and it will be a deadlock

Reply:
@hangyujin I just tried the above code in a dummy project and didn't get a deadlock. Note that im calling from `viewDidLoad`, but the idea should be similar. 

```
@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [self fakeRunloop];
  NSLog(@"completed fakeRunLoop");
}
@end
```

As you can see `NSLog(@"completed fakeRunLoop");` gets printed out. 

![Screenshot 2024-05-23 at 1 05 19 PM](https://github.com/flutter/engine/assets/41930132/6b5243f6-b40c-43bd-bc12-579d52e5c646)

What exact behavior are you observing with this code?

Reply:
I called it from  application:continueUserActivity:restorationHandler because this method will be invoked by deep linking, 

```
- (BOOL)application:(UIApplication*)application
    continueUserActivity:(NSUserActivity*)userActivity
      restorationHandler:
          (void (^)(NSArray<id<UIUserActivityRestoring>>* __nullable restorableObjects))
              restorationHandler {

  NSLog(@"application:continueUserActivity:restorationHandler"); 
  [self fakeRunloop];
  NSLog(@"completed fakeRunLoop");
  return NO;
}
```
And it's just running forever and can't complete 
<img width="880" alt="Screenshot 2024-05-23 at 14 32 32" src="https://github.com/flutter/engine/assets/108393416/bfe06044-7f88-4cbb-bcfa-f006c7b9dfa0">

Reply:
I will look into this today

Reply:
Just wanna update my progress so far. So for the deadlock, I wasn't able to find anything useful online, but I did come up with a theory - the `continueUserActivity` runs on main queue (hence also the main thread); the `viewDidLoad` runs on main thread, but NOT the main queue. This can explain why it works in `viewDidLoad`, but deadlock in `continueUserActivity`. 

If my theory is correct, I don't think there's a way to synchronize the async API. I am thinking about a workaround (which I am still hashing out). Basically we want to return `YES` synchronously (marking as "handled"), then call the async API. If the async API comes back as failure, then call `UIAppilcation::openURL` which should throw the url back to iOS. Something like: 

```
- (BOOL) continueUserActivity {
  if vc not present { return NO; }
  runAsyncAPI(completion: (success) { 
    if (!success) { 
      // throw it back to iOS
      [UIApplication.sharedApplication openURL: url]
    }
  })
  return YES;
}
```

Reply:
UIAppilcation::openURL which should throw the url back to iOS >> what will happen then? will it open the url in the browser?

Reply:
@hangyujin My understanding is that, it should open in the next best candidate app that registers the domain, and browser is the catch-all default.

Reply:
I am still playing around with the deep link API, but from my personal experience with Amazon: 

Whenever i get an amazon package, I receive a link through text that starts with `https://a.co/d/WHATEVER_CODE`. When I click on it, it launches my Amazon app (which means amazon app registers `https://a.co`), and then after a few seconds, it redirect to Safari that displays the bar code. 

This makes me think that Amazon app probably returned YES in `continueUserActivity`, asynchronously run some code that determines that the app can't handle it, so it throws the url back to iOS, which opens Safari as the next best candidate.

Reply:
I updated the code to use [UIApplication.sharedApplication openURL: url] to throw it back to iOS.
Also tested that custom URI scheme is using a different API application:openURL:options 
I also tried return NO in application:continueUserActivity:restorationHandler for (universal link)and  application:openURL:options (for custom URI), return NO will not cause it to throw the URL back, the behavior looks the same as return YES.

Also `go_router` package  have a fallback error screen if the deep link path is not set in the flutter package. So if you’re using go_router for deep linking, it will always consider the deep link handled and the `throw back to iOS` thing won’t happen. 
![unnamed-1](https://github.com/flutter/engine/assets/108393416/60d85977-2cc7-4591-bf6c-93826cf69247)